### DISCUSSION TITLE
Serialization and MobileFormatter futures

### DISCUSSION THREAD

**Discussion by rockfordlhotka** [MarimerLLC/csla#2386] [Ideas]
This is a thread for discussing possible futures around serialization and/or MobileFormatter as we look forward.

As this thread starts there are several related backlog items already listed, but this thread isn't limited by them.

* https://github.com/MarimerLLC/csla/issues/1297 (done: @rockfordlhotka)
* https://github.com/MarimerLLC/csla/issues/2531
* https://github.com/MarimerLLC/csla/issues/2346
* https://github.com/MarimerLLC/csla/issues/2154
* https://github.com/MarimerLLC/csla/issues/2148 (active: @TheCakeMonster)
* https://github.com/MarimerLLC/csla/issues/1877
* https://github.com/MarimerLLC/csla/issues/1628
* https://github.com/MarimerLLC/csla/issues/1376
* https://github.com/MarimerLLC/csla/issues/1131
* https://github.com/MarimerLLC/csla/issues/758

Please keep in mind that CSLA 5 is "done" and all work is now occurring for CSLA 6 or later.

**Comment by rockfordlhotka** [comment]
From another thread @bradtwurst says:

> As I am attempting to migrate away from WCF and deal with a lot of POCO objects, I found my way here.
> 
> I found MobileFormatter very tough (frustrating) to deal with as it is sealed, locked-down, and is currently hard-coded in many of the portal implementations (although I do see that this last issue has changed in main branch).
> 
> While this may not be the ultimate solution, would it make sense to have an OnSerializing/Deserializing method hooks that could be utilized to allow the app developer to inherit from MF and tweak the logic.
> 
> I know that this wouldn't solve all the things, but it is a fairly straight-forward pattern that was used in other parts of the framework and it could be implemented in the current v5 release without breaking all the things.
> 
> If it sounds like an acceptable strategy, I'd be willing to invest some effort into this area as it is directly affecting my current focus.
> 
> James

One of the things @TheCakeMonster and I have been discussing over the past few weeks, is the idea of enhancing MF so it accepts a list of custom serializer types, where a serializer type knows how to serialize another type.

A scenario where this would be useful is a type that knows how to serialize POCO types, so you'd register it with MF and it would "somehow" identify which types it is willing to serialize (maybe via a custom attribute on the type or whatever).

Serializing _arbitrary_ types is a potential security risk, which is why Microsoft has abandoned BinaryFormatter and NDCS. MF itself can only serialize types that opt-in to the process - you must code your type to participate in serialization, so the risk is not the same as BF/NDCS. I'd expect the same from any POCO serializer we build into the framework.

However, this ability to implement a custom type serializer would allow _you_ to implement whatever you'd like in your own codebase - assuming your own risk.

You can imagine that a custom type serializer might implement an interface something like:

```c#
public interface ISerializeType
{
  bool CanSerialize(Type type);
  void Serialize(SerializationIinfo info, object target);
  object Deserialize(SerializationInfo info);
}
```

That's off the type of my head, but is the gist of the idea. MF itself would, for types it doesn't already understand, loop through the registered type serializers asking each whether it knows how to serialize the current type. If the answer is yes, MF will delegate serialization/deserialization to the type serializer. If the answer is no, MF will keep checking and eventually (if no match is found) throw the same serialization exception it does today.

**Reply by rockfordlhotka** [reply]
Another thing @TheCakeMonster and I discussed is a more directed solution just for POCOs, which is to have MF itself look for a custom attribute on a type, and if the type has that attribute, MF will serialize _all fields of any scope_ in the type, basically just like BF does today.

That is very easy for a single object. It is a lot more complex for an _object graph_, because deserialization of an object graph is ... challenging. It requires doing what BF, NDCS, MF, and other full fidelity serializers do, which is to flatten the graph, set up instance ids, and serialize _the flat graph_, then use the instance ids to recreate the graph during deserialization.

But totally doable within the context of MF, because it does the heavy lifting.

**Reply by TheCakeMonster** [reply]
It sort of feels to me like the interface that the o/p was asking for already exists - isn't this something you could mostly/almost completely do by implementing IMobileObject?

In past I was reticent to implement this interface, thinking it was very complex and I'd get it wrong. In fact, when it came time to do my work on the source generator for serializing POCOs, I came to understand it better, and realised that my fears were unfounded. 

I wonder if, with a little more documentation or guidance, we might be able to make the IMobileObject interface much more approachable, and avoid the need for extra complexity in the framework.

**Reply by kcabral817** [reply]
There are a couple of types from MS Sql Server that might be considered for mobile formatter upgrade.  I know of three at this time, DateTimeOffset, Geography and Geometry.

**Reply by rockfordlhotka** [reply]
The core Csla project won't take a dependency on any type that isn't in .NET itself - so no types from things like `Microsoft.Data.SqlClient`, etc.

What @TheCakeMonster and I have been discussing though, is a modification to `MobileFormatter` where _you_ can provide a serializer for specific types. You'd register these serializers during app startup (via DI), and when `MobileFomatter` encounters one of these types, it would delegate the value to your serializer to convert it to/from a string (or byte array).

**Reply by TheCakeMonster** [reply]
Ah, yes. An excellent use case for the capability we were discussing. Good to have other real-world examples of what we are trying to add support for, thanks.

**Reply by rockfordlhotka** [reply]
I think that if we have `MobileFormatter` rely on a dictionary of <targettype, serializertype> that can be defined during app startup, and the dictionary injected to `MobileFormatter` (or something) that we'd have a good/flexible solution.

Then, if we want to use an attribute-based model, the attribute can be used to help load the dictionary, but _also_, types that can't be decorated with the attribute can be directly added to the dictionary during app startup.

**Reply by rockfordlhotka** [reply]
https://github.com/MarimerLLC/csla/issues/2531
